rllib/utils/replay_buffers/storage.py

```diff
@@ -0,0 +1,503 @@
+        """
+        if idx < 0:
+            raise IndexError('Buffer index out of range')
+        return (self._offset_idx + idx) % self.capacity
+
+    def _get_external_index(self, idx: int):
+        """Translate the given internal circular buffer index into
+        the external index space of the storage.
+
+        Args:
+            idx: Internal circular Buffer index (0 <= idx < capacity).
+
+        Returns:
+            External index from interval [0, len(storage))
+        """
+        if idx < 0:
+            raise IndexError('Buffer index out of range')
+        if idx >= self._offset_idx:
+            return idx - self._offset_idx
+        else:
+            return idx + self.capacity - self._offset_idx
+
+    @abstractmethod
+    def _get(self, idx: int) -> SampleBatchType:
+        """Get the item at the specified index / key.
+
+        This method must be implementend by subclasses
+        using an actual data structure for storing the data.
+        This data structure must be capable of dealing with
+        indices between 0 <= idx < capacity.
+
+        Args:
+            idx: Index of the item of interest.
+
+        Returns:
+            Item at index.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def _set(self, idx: int, item: SampleBatchType) -> None:
+        """Store the given item at the specified index / key.
+
+        This method must be implementend by subclasses
+        using an actual data structure for storing the data.
+        This data structure must be capable of dealing with
+        indices between 0 <= idx < capacity.
+
+        Args:
+            idx: Index to store the item at.
+            item: Item to store at specified index.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def _del(self, idx: int) -> SampleBatchType:
+        """Remove and return the item at the specified index / key.
+
+        This method must be implementend by subclasses
+        using an actual data structure for storing the data.
+        This data structure must be capable of dealing with
+        indices between 0 <= idx < capacity.
+
+        Args:
+            idx: Index of the item of interest.
+
+        Returns:
+            Item at index that has been removed.
+        """
+        raise NotImplementedError()
+
+
+@ExperimentalAPI
+class InMemoryStorage(LocalStorage):
+
+    @ExperimentalAPI
+    @override(LocalStorage)
+    def __init__(self, capacity: int = 10000) -> None:
+        """Initializes an InMemoryStorage instance for storing timesteps in memory.
+
+        The storage uses a numpy array as datastructure.
+
+        Args:
+            capacity: Max number of timesteps to store in this FIFO
+                buffer. After reaching this number, older samples will be
+                dropped to make space for new ones.
+        """
+        super().__init__(capacity)
+        self._samples = np.empty((self._capacity,), dtype=object)
```

Comment:
Just a comment: I expected this to be quicker for a capacity of 1M, but timeit tells me that it's the same as list.

Reply:
Cool, I did not run any performance tests yet. I assume that the performance did change in some situations.